packages/core/src/tools/sanitize.ts

```diff
@@ -0,0 +1,235 @@
+        ;(containerToProcess.target as ContextArray)[key] = targetData
+      }
+    } else {
+      for (const key in containerToProcess.source) {
+        if (Object.prototype.hasOwnProperty.call(containerToProcess.source, key)) {
+          const targetData = sanitizeProcessor(
+            containerToProcess.source[key],
+            containerToProcess.path,
+            key,
+            containerQueue,
+            visitedObjectsWithPath
+          )
+          accumulatedSize += computeSize(targetData, separatorLength, key)
+          if (accumulatedSize > maxCharacterCount) {
+            break
+          }
+          separatorLength = 1
+          ;(containerToProcess.target as Context)[key] = targetData
+        }
+      }
+    }
+  }
+
+  // Rebind detached toJSON functions
+  restoreObjectPrototypeToJson()
+  restoreArrayPrototypeToJson()
+
+  return sanitizedData
+}
+
+/**
+ * Internal function to factorize the process common to the
+ * initial call to sanitize, and iterations for Arrays and Objects
+ *
+ */
+function sanitizeProcessor(
+  source: ExtendedContextValue,
+  parentPath: string,
+  key: string | number | undefined,
+  queue: ContainerElementsToProcess[],
+  visitedObjectsWithPath: WeakMap<object, string>
+) {
+  // Start by handling toJSON, as we want to sanitize its output
+  const sourceToSanitize = tryToApplyToJSON(source)
+
+  if (!sourceToSanitize || typeof sourceToSanitize !== 'object') {
+    return sanitizePrimitivesAndFunctions(sourceToSanitize)
+  }
+
+  const sanitizedSource = sanitizeObjects(sourceToSanitize)
+  if (sanitizedSource !== '[Object]' && sanitizedSource !== '[Array]') {
+    return sanitizedSource
+  }
+
+  // Handle potential cyclic references
+  // We need to use source as sourceToSanitize could be a reference to a new object
+  // At this stage, we know the source is an object type
+  const sourceAsObject = source as object
+  if (visitedObjectsWithPath.has(sourceAsObject)) {
+    return `[Reference seen at ${visitedObjectsWithPath.get(sourceAsObject)!}]`
+  }
+
+  // Add processed source to queue
+  const currentPath = key ? `${parentPath}.${key}` : `${parentPath}`
+  const target = Array.isArray(sourceToSanitize) ? [] : {}
+  visitedObjectsWithPath.set(sourceAsObject, currentPath)
+  queue.push({ source: sourceToSanitize, target, path: currentPath })
+
+  return target
+}
+
+/**
+ * Used to approximate the number of characters elements occupy inside a container
+ * "key":"value" => returns a size of 13 if first value in collection (separatorLength=0), 14 otherwise (separatorLength=1)
+ *
+ * LIMITATIONS
+ * - Actual byte count may differ according to character encoding
+ *
+ * @param value              value in key/value pair
+ * @param separatorLength    0 if first pair in container, 1 otherwise (accounts for the comma separator)
+ * @param key                key in key/value pair (undefined for arrays)
+ *
+ */
+function computeSize(value: ContextValue, separatorLength = 0, key?: string) {
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏I am having a hard time following what is done with size computation regarding separator and key, I am wondering if it could not be easier by making things a bit more explicit in each case rather than having this methods handling all the cases.

Reply:
👌